### ISSUE TITLE
False-positive protobuf-undefined-attribute, nested messages variable re-assignment

### ISSUE THREAD

**Issue by zoedberg** [nelfin/pylint-protobuf#44]
using this `example.proto`:
```python
syntax = "proto3";
message Example {
  int32 value = 1;
  float fraction = 2;
}
message DifferentExample {
  int32 different_value = 1;
  float different_fraction = 2;
}
```
and linting this python script:
```python
from example_pb2 import DifferentExample, Example

request = Example(value=6)
if 1 + 1 == 2:
    request = DifferentExample()
    if 7 + 1 == 8:
        request.different_value = 9
```

I receive the following false positive error: `E5901: Field 'different_value' does not appear in the declared fields of protobuf-generated class 'Example' and will raise AttributeError on access (protobuf-undefined-attribute)`

The error seems related to the re-assignment of the `request` variable and the double nesting.

**Comment by nelfin** [comment]
Hi @zoedberg, thanks for the report. When inferring the value of `request` at line 7, `astroid` doesn't attempt to resolve the ambiguous case, see the following example:

```python
import astroid

node = astroid.extract_node("""
    items = [1, 2, 3]
    val = 1
    if 1 + 1 == 2:
        val = 'blue'
    val
""")
print(list(node.infer()))
# >>> [<Const.int l.3 at 0x...>, <Const.str l.5 at 0x...>]
```

For built-in checkers like `pylint.checkers.typecheck` they resolve this by checking if any object, say, has the missing member and only raising the warning if *none* of them do, see this paraphrased example from `pylint.checkers.typecheck:visit_attribute`:

```python
missingattr = set()
for owner in inference_results:
    try:
        if not owner.getattr(node.attrname):
            missingattr.add((owner, name))
            continue
    except exceptions.NotFoundError:
        if not _emit_no_member(node, owner, name):
            continue
        missingattr.add((owner, name))
        continue
    break # stop on the first found
else:
    # we have not found any node with the attributes, display the
    # message for inferred nodes
    self.add_message(...)
```

That `for`-`else` clause will only branch if the loop is not broken out of (by one of the results actually passing the membership test). In `pylint_protobuf.visit_attribute` we only checked the first inference result that looked like a protobuf class.

I'll see about updating the logic there to match the pylint behaviour.

**Comment by nelfin** [comment]
Note that this does lead to *false-negatives*, such as the following:

```python
class Example:
    one = 'red'

class DifferentExample:
    two = 'blue'

ex = Example()
if 1 + 1 == 2:
    ex = DifferentExample()
    if 2 + 2 == 4:
        print(ex.one)
```

which does not raise `no-member`. I assume that false-negatives are fine.

**Comment by nelfin** [comment]
Can you see if the latest version `pip install git+https://github.com/nelfin/pylint-protobuf@2c09cf32a11f5f14ec27dd2ee1c8daa307c4ea38` addresses your issue?

**Comment by zoedberg** [comment]
> Can you see if the latest version pip install git+https://github.com/nelfin/pylint-protobuf@2c09cf32a11f5f14ec27dd2ee1c8daa307c4ea38 addresses your issue?

it addresses the example script I've provided before, but it doesn't work when proto messages are imported like this:
```python
import example_pb2 as pb

request = pb.Example(value=6)
if 1 + 1 == 2:
    request = pb.DifferentExample()
    if 7 + 1 == 8:
        request.different_value = 9
```

**Comment by zoedberg** [comment]
>  I assume that false-negatives are fine.

Not sure about this. False-positive are annoying (but one can just ignore them knowing the linter is incorrect), but on the contrary false-negatives can be dangerous because they could hide actual issues

**Comment by nelfin** [comment]
> False-positive are annoying (but one can just ignore them knowing the linter is incorrect), but on the contrary false-negatives can be dangerous because they could hide actual issues

False-positives lead to fatigue and people just ignoring the linter entirely, not only the messages
that they know are incorrect. Pylint has been routinely critised in the past for having too high
a false-positive rate (it's tagline was "It’s not just a linter that annoys you!"). False-negatives,
while reducing the utility of the linter, leave you no better off than you were before you decided
to check your code. A linter should not be a mission-critical part of your software, it can only
give you code quality suggestions and thus should be part of a wide variety of techniques for ensuring
correctness and fault-tolerance: robust testing, defensive programming, code review, etc.

There will presumably always be some (low!) level of acceptable false-negative rate. If you feel that
it's too high then we can work to try and improve it, but it's a "best effort" kind of thing, not
any guarantee.

> it addresses the example script I've provided before, but it doesn't work when proto messages are imported like this

I'll have a look at what's causing the problems with your second example.

**Comment by nelfin** [comment]
What version of `astroid` and `pylint` are you running? Strangely, I couldn't replicate this in my environment with `astroid==2.5` and `pylint==2.6.0` but I tried again in a fresh environment and saw the warning that you talked about in https://github.com/nelfin/pylint-protobuf/issues/44#issuecomment-808892079 with `astroid==2.5.2` and `pylint==2.7.2`.

Here's the script I used to replicate this:

```sh
#!/bin/sh

set -e

[ -d "issue44" ] && rm -rf issue44
mkdir issue44 && cd issue44

python -m venv venv
. venv/bin/activate

cat > requirements.txt <<EOF
# astroid==2.5.2
astroid==2.5
isort==5.8.0
lazy-object-proxy==1.6.0
mccabe==0.6.1
protobuf==3.15.6
# pylint==2.7.2
pylint==2.6.0
git+https://github.com/nelfin/pylint-protobuf@2c09cf32a11f5f14ec27dd2ee1c8daa307c4ea38
six==1.15.0
toml==0.10.2
typed-ast==1.4.2
wrapt==1.12.1
EOF
pip install -r requirements.txt

cat > example.proto <<EOF
syntax = "proto3";
message Example {
  int32 value = 1;
  float fraction = 2;
}
message DifferentExample {
  int32 different_value = 1;
  float different_fraction = 2;
}
EOF
protoc --python_out=. example.proto

cat > issue44b.py <<EOF
import example_pb2 as pb

request = pb.Example(value=6)
if 1 + 1 == 2:
    request = pb.DifferentExample()
    if 7 + 1 == 8:
        request.different_value = 9
EOF

pylint --load-plugins=pylint_protobuf issue44b.py
```

With `astroid==2.5` and `pylint==2.6.0`:
```
************* Module issue44b
issue44b.py:1:0: C0114: Missing module docstring (missing-module-docstring)

------------------------------------------------------------------
Your code has been rated at 8.33/10 (previous run: 0.00/10, +8.33)
```

With `astroid==2.5.2` and `pylint==2.7.2`:
```
************* Module issue44b
issue44b.py:1:0: C0114: Missing module docstring (missing-module-docstring)
issue44b.py:7:8: E5901: Field 'different_value' does not appear in the declared fields of protobuf-generated class 'Example' (protobuf-undefined-attribute)

------------------------------------------------------------------
Your code has been rated at 0.00/10 (previous run: 8.33/10, -8.33)
```

**Comment by nelfin** [comment]
Note that this is not `pylint_protobuf` specific, I've changed the behaviour to not attempt checking attributes if one of the branches is uninferable.

**Comment by nelfin** [comment]
CC @zoedberg: I've raised https://github.com/PyCQA/astroid/issues/926 upstream for this, and I think I may have found a fix

**Comment by zoedberg** [comment]
> What version of astroid and pylint are you running?

Not sure you still need this information, but I'm running `astroid==2.5.1` and `pylint==2.7.2`

> There will presumably always be some (low!) level of acceptable false-negative rate. If you feel that
it's too high then we can work to try and improve it, but it's a "best effort" kind of thing, not
any guarantee.

I don't feel it's too high, but if one day I'll have some spare time I could look into this, I would be happy to help

**Comment by nelfin** [comment]
This has been released in `v0.20.0` just now
